Lib/importlib/_bootstrap_external.py

```diff
@@ -312,7 +322,33 @@ def cache_from_source(path, debug_override=None, *, optimization=None):
         if not optimization.isalnum():
             raise ValueError('{!r} is not alphanumeric'.format(optimization))
         almost_filename = '{}.{}{}'.format(almost_filename, _OPT, optimization)
-    return _path_join(head, _PYCACHE, almost_filename + BYTECODE_SUFFIXES[0])
+    filename = almost_filename + BYTECODE_SUFFIXES[0]
+    if sys.bytecode_path is not None:
+        # We need an absolute path to the py file to avoid the possibility of
+        # collisions within sys.bytecode_path, if someone has two different
+        # `foo/bar.py` on their system and they import both of them using the
+        # same sys.bytecode_path. Let's say sys.bytecode_path is
+        # `C:\Bytecode`; the idea here is that if we get `Foo\Bar`, we first
+        # make it absolute (`C:\Somewhere\Foo\Bar`), then make it root-relative
+        # (`Somewhere\Foo\Bar`), so we end up placing the bytecode file in
+        # an unambiguous `C:\Bytecode\Somewhere\Foo\Bar\`.
+        if not _path_isabs(head):
+            head = _path_join(_os.getcwd(), head)
```

Comment:
If the import system is ever handling a non-absolute path, AFAIK "relative to CWD" is the existing (and only reasonable) interpretation for what that path means. E.g. the empty string in `sys.path` means CWD.

Reply:
Sorry, my memory failed me I guess (no surprise there).  I could swear we resolve that first sys.path entry at startup.  I'm probably thinking of a proposed change (though I couldn't find a bpo issue or python-ideas thread).  Following the existing convention (e.g. with sys.path) is definitely the right thing to do.  Carry on. :)

Reply:
@ncoghlan, do you recall any such proposal?  My expectation/intuition was that the import system would not be affected by changing CWD (and I'd expect most folks to have the same expectation).  The alternative (apparently the status quo) seems like something that would lead to hard-to-diagnose bugs in some corner cases.  If no such proposal exists then I'll probably make one. :)

Reply:
@ericsnowcurrently You may be thinking of https://bugs.python.org/issue33053, where we changed the `-m` switch to add the starting directory instead of the current directory.

The `-c` switch, the interactive prompt, and non-interactive `stdin` execution still use the empty string as `sys.path[0]`, though, even in Python 3.7. (There's a reasonable case to be made that all non-interactive modes should use the starting directory, but things are significantly less clear when it comes to the interactive prompt, since different folks use that in very different ways, and the current behaviour has been around for a *long* time)

Reply:
Yeah, I'm pretty sure that's the one.  I'll follow up separately.